src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1916 @@
+                    var token = child.Token;
+                    foreach (var trivia in token.LeadingTrivia)
+                    {
+                        AddUniqueDiagnostics(trivia.Diagnostics, diagnostics);
+                    }
+
+                    AddUniqueDiagnostics(token.Diagnostics, diagnostics);
+                }
+            }
+        }
+
+        /// <summary>
+        /// It's very common to have duplicated diagnostics.  For example, consider "((". This will
+        /// have two 'missing )' diagnostics, both at the end.  Reporting both isn't helpful, so we
+        /// filter duplicates out here.
+        /// </summary>
+        private static void AddUniqueDiagnostics(ImmutableArray<EmbeddedDiagnostic> from, ArrayBuilder<EmbeddedDiagnostic> to)
+        {
+            foreach (var diagnostic in from)
+            {
+                if (!to.Contains(diagnostic))
+                {
+                    to.Add(diagnostic);
+                }
+            }
+        }
+
+        private RegexExpressionNode ParseAlternatingSequences(bool consumeCloseParen)
+        {
+            try
+            {
+                _recursionDepth++;
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
+                return ParseAlternatingSequencesWorker(consumeCloseParen);
+            }
+            finally
+            {
+                _recursionDepth--;
+            }
+        }
+
+        /// <summary>
+        /// Parses out code of the form: ...|...|...
+        /// This is the type of code you have at the top level of a regex, or inside any grouping
+        /// contruct.  Note that sequences can be empty in .net regex.  i.e. the following is legal:
+        /// 
+        ///     ...||...
+        /// 
+        /// An empty sequence just means "match at every position in the test string".
+        /// </summary>
+        private RegexExpressionNode ParseAlternatingSequencesWorker(bool consumeCloseParen)
+        {
+            RegexExpressionNode current = ParseSequence(consumeCloseParen);
+
+            while (_currentToken.Kind == RegexKind.BarToken)
+            {
+                // Trivia allowed between the | and the next token.
+                current = new RegexAlternationNode(
+                    current, ConsumeCurrentToken(allowTrivia: true), ParseSequence(consumeCloseParen));
+            }
+
+            return current;
+        }
+
+        private RegexSequenceNode ParseSequence(bool consumeCloseParen)
+        {
+            var list = ArrayBuilder<RegexExpressionNode>.GetInstance();
+
+            if (ShouldConsumeSequenceElement(consumeCloseParen))
+            {
+                do
+                {
+                    var last = list.Count == 0 ? null : list.Last();
+                    list.Add(ParsePrimaryExpressionAndQuantifiers(last));
+
+                    TryMergeLastTwoNodes(list);
+                }
+                while (ShouldConsumeSequenceElement(consumeCloseParen));
```

Comment:
Perhaps make `shouldConsumeSequenceElement` a local function.

Reply:
I'm ok as-is.  #won'tfix